extensions/ql-vscode/src/language-context-store.ts

```diff
@@ -0,0 +1,53 @@
+import { App } from "./common/app";
+import { DisposableObject } from "./common/disposable-object";
+import { AppEvent, AppEventEmitter } from "./common/events";
+import { QueryLanguage } from "./common/query-language";
+
+type LanguageFilter = QueryLanguage | "All";
+
+export class LanguageContextStore extends DisposableObject {
+  public readonly onLanguageContextChanged: AppEvent<void>;
+  private readonly onLanguageContextChangedEmitter: AppEventEmitter<void>;
+
+  private state: LanguageFilter;
+
+  constructor(private readonly app: App) {
+    super();
+    // State initialization
+    this.state = "All";
+
+    // Set up event emitters
+    this.onLanguageContextChangedEmitter = this.push(
+      app.createEventEmitter<void>(),
+    );
+    this.onLanguageContextChanged = this.onLanguageContextChangedEmitter.event;
+  }
+
+  public async clearLanguageContext() {
+    this.state = "All";
+    this.onLanguageContextChangedEmitter.fire();
+    await this.app.commands.execute(
+      "setContext",
+      "codeQLDatabases.languageFilter",
+      "",
+    );
+  }
+
+  public async setLanguageContext(language: QueryLanguage) {
+    this.state = language;
+    this.onLanguageContextChangedEmitter.fire();
+    await this.app.commands.execute(
+      "setContext",
+      "codeQLDatabases.languageFilter",
+      language,
+    );
+  }
+
+  // This method takes a string to allow it to be used in cases
+  // where the language is not always a known one.
+  // The semantics of such an unknown langauge is that it is
+  // only included if the current language context is "All".
+  public shouldInclude(language: string): boolean {
```

Comment:
Can we use [`isQueryLanguage`](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/src/common/query-language.ts#L62) here to help us out?

Reply:
Not sure what we would use it for. We could check that the string is a query language, but because the only thing that can be put into the state is a query language or all, then the check inside the second clause of the disjunction should amount to the same thing. But I might be missing something?

Reply:
We've made the argument of this function a `string` because the caller could have a `string` rather than a `QueryLanguage` (if I'm understanding this right). But only valid QueryLanguage values should be allowed. So it'd be good for the argument to change to a `QueryLanguage` and the caller to user `isQueryLanguage` to make sure it's sending the right thing.

Reply:
The trouble with having the caller use `isQueryLanguage` is that makes it hard to preserve the semantics (that unknown values should be shown when "All" is selected) without leaking "All" outside the store, or having to expose a method like "IsAllSelected".

One alternative is to pass `QueryLanguage | undefined` (because I actually need that for the queries panel). However that does make it more annoying for the callers, as they then have to convert their `string` to a `QueryLanguge | undefined`. But that is not the end of the world, and we could do that as a helper method similar to `isQueryLanguage`.

The last alternative is to expose `LanguageFilter` outside the store, which is also fine but it seemed nice to encapsulate the "All" implementation.

Reply:
`QueryLanguage | undefined` sounds like it makes the most sense to me (and I'm okay with the callers having to put some extra effort in) but I'm also happy with whatever you decide - I've not thought about this deeply, and it's easily changeable if we find that it's causing problems too.

Reply:
Pushed a new version that takes in `QueryLanguage | undefined`.